# Libraries API Guide
[Documentation](https://github.com/Kotlin/kotlin-jupyter/blob/master/docs/libraries.md)

In [2]:
notebook.kernelVersion

0.12.0.128

## Dependencies and imports

Assume we want to add binary dependency on some library.
In this case let's add JAR coordinates of this library to the `dependencies`
array of our descriptor. If it is not on Maven Central, we add repository
(or repositories) where this dependency is located. Also, we can add
default imports

In [3]:
val dollar = "$"
loadLibraryDescriptor(
    """
    {
        "dependencies": [
            "org.ileasile:kformat:${dollar}libVersion"
        ],
        "repositories": [
            "https://maven.pkg.github.com/ileasile/kformat"
        ],
        "imports": [
            "org.ileasile.kformat.*"
        ],
        "properties": {
            "libVersion": "0.0.1"
        }
    }
    """,
    mapOf("libVersion" to "0.0.2")
)

The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
Error compiling code:
@file:Repository("https://maven.pkg.github.com/ileasile/kformat")
@file:DependsOn("org.ileasile:kformat:0.0.2")
import org.ileasile.kformat.*


Errors:
Failed to resolve [org.ileasile:kformat:0.0.2]:
File 'org.ileasile:kformat:0.0.2' not found
ArtifactResolutionException: Could not transfer artifact org.ileasile:kformat:pom:0.0.2 from/to https___maven.pkg.github.com_ileasile_kformat (https://maven.pkg.github.com/ileasile/kformat): authentication failed for https://maven.pkg.github.com/ileasile/kformat/org/ileasile/kformat/0.0.2/kformat-0.0.2.pom, status: 401 Unauthorized

org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryException: The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
	at org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryExceptionKt.rethrowAsLibraryException(ReplLibraryException.kt:32)
	a

Huh, we've got 401(Unauthorized). That's because reading from GitHub Packages
requires authorization. We can add authorization for our repository in descriptor.

In [4]:
val readToken = "ghp_<your read token>"

In [5]:
loadLibraryDescriptor("""
    {
        "dependencies": [
            "org.ileasile:kformat:0.0.2"
        ],
        "repositories": [
            {
                "path": "https://maven.pkg.github.com/ileasile/kformat",
                "username": "ileasile",
                "password": "$readToken"
            }
        ],
        "imports": [
            "org.ileasile.kformat.*"
        ]
    }
""")

The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
Error compiling code:
@file:Repository("https://maven.pkg.github.com/ileasile/kformat", username="ileasile", password="ghp_<your read token>")
@file:DependsOn("org.ileasile:kformat:0.0.2")
import org.ileasile.kformat.*


Errors:
Failed to resolve [org.ileasile:kformat:0.0.2]:
File 'org.ileasile:kformat:0.0.2' not found
ArtifactResolutionException: Could not transfer artifact org.ileasile:kformat:pom:0.0.2 from/to https___maven.pkg.github.com_ileasile_kformat (https://maven.pkg.github.com/ileasile/kformat): authentication failed for https://maven.pkg.github.com/ileasile/kformat/org/ileasile/kformat/0.0.2/kformat-0.0.2.pom, status: 401 Unauthorized

org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryException: The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
	at org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryExceptionKt.r

In [6]:
kFormat {
    color(BasicColor.Red) {
        italic {
            + "Red italic text"
        }
    }
}.asAnsiEscaped()

Line_7.jupyter.kts (1:1 - 8) Unresolved reference: kFormat
Line_7.jupyter.kts (2:5 - 10) Unresolved reference: color
Line_7.jupyter.kts (2:11 - 21) Unresolved reference: BasicColor
Line_7.jupyter.kts (3:9 - 15) Unresolved reference: italic
Line_7.jupyter.kts (4:13 - 14) Unresolved reference: +

Another way to do the same thing using Kotlin API:

In [7]:
USE {
    dependencies("org.ileasile:kformat:0.0.2")
    repository(
        "https://maven.pkg.github.com/ileasile/kformat",
        "ileasile",
        readToken
    )
    import("org.ileasile.kformat.*")
}

The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
Error compiling code:
@file:Repository("https://maven.pkg.github.com/ileasile/kformat", username="ileasile", password="ghp_<your read token>")
@file:DependsOn("org.ileasile:kformat:0.0.2")
import org.ileasile.kformat.*


Errors:
Failed to resolve [org.ileasile:kformat:0.0.2]:
File 'org.ileasile:kformat:0.0.2' not found
ArtifactResolutionException: Could not transfer artifact org.ileasile:kformat:pom:0.0.2 from/to https___maven.pkg.github.com_ileasile_kformat (https://maven.pkg.github.com/ileasile/kformat): authentication failed for https://maven.pkg.github.com/ileasile/kformat/org/ileasile/kformat/0.0.2/kformat-0.0.2.pom, status: 401 Unauthorized

org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryException: The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
	at org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryExceptionKt.r

Also, Kotlin API allows to define dependencies in Gradle style:

In [8]:
USE {
    dependencies {
        implementation("org.ileasile:kformat:0.0.2")
    }
    repositories {
        maven {
            url = "https://maven.pkg.github.com/ileasile/kformat"
            credentials {
                username = "ileasile"
                password = readToken
            }
        }
    }
    import("org.ileasile.kformat.*")
}

The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
Error compiling code:
@file:Repository("https://maven.pkg.github.com/ileasile/kformat", username="ileasile", password="ghp_<your read token>")
@file:DependsOn("org.ileasile:kformat:0.0.2")
import org.ileasile.kformat.*


Errors:
Failed to resolve [org.ileasile:kformat:0.0.2]:
File 'org.ileasile:kformat:0.0.2' not found
ArtifactResolutionException: Could not transfer artifact org.ileasile:kformat:pom:0.0.2 from/to https___maven.pkg.github.com_ileasile_kformat (https://maven.pkg.github.com/ileasile/kformat): authentication failed for https://maven.pkg.github.com/ileasile/kformat/org/ileasile/kformat/0.0.2/kformat-0.0.2.pom, status: 401 Unauthorized

org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryException: The problem is found in one of the loaded libraries: check library imports, dependencies and repositories
	at org.jetbrains.kotlinx.jupyter.exceptions.ReplLibraryExceptionKt.r

In [9]:
//@file:DependsOn
//@file:Repository

## Initialization callbacks
You can define the code that should be run after library initialization.
In descriptors, only code executions are available for this purpose.
In Kotlin API you can do a bit more.

In [10]:
loadLibraryDescriptor("""
    {
        "init": [
            "val myVar1 = 42",
            "fun doInit1() {}",
            "println(\"Library 1 initialized\")"
        ]
    }
""")

Library 1 initialized


In [11]:
class Config(
    var bufferSize: Int = 100
)

val config1 = Config()

USE {
    onLoaded {
        declare("myVar3" to config1)
    }
}

In [12]:
myVar3.bufferSize = myVar1

In [13]:
myVar3.bufferSize

42

In [14]:
myVar3.bufferSize = 10

In [15]:
config1.bufferSize

10

## Before- and after-cell callbacks

Sometimes you may want to execute some code before cell execution.
That's how you can achieve it with library descriptor

In [16]:
loadLibraryDescriptor("""
    {
        "initCell": [
            "println(\"Before cell execution - 1\")"
        ]
    }
""")

In [17]:
2 + 2 + 7

Before cell execution - 1


11

You can achieve the same thing using Kotlin API

In [18]:
USE {
    beforeCellExecution {
        println("Before cell execution - 2")
    }
}

Before cell execution - 1


In [19]:
2 + 2

Before cell execution - 1
Before cell execution - 2


4

You can manage added callbacks. Let's clear them.

In [20]:
notebook.beforeCellExecutionsProcessor.unregisterAll()

Before cell execution - 1
Before cell execution - 2


In [21]:
2 + 2

4

With Kotlin API, you can also create callbacks that are run after cells execution. Execution result is passed to them.

In [22]:
USE {
    afterCellExecution { snippetInstance, result ->
        val maybeDoubledResult = (result.value as? Int)?.let { it * 2 }
        println("After execution of $snippetInstance: $maybeDoubledResult (${result.name})")
    }
}

In [23]:
2 + 2

After execution of Line_35_jupyter@149fe128: 8 (res31)


4

Let's remove this callback as well, but do it in a bit more complicated manner

In [24]:
notebook.afterCellExecutionsProcessor.let { processor ->
    processor.unregister(processor.registeredExtensions().first())
}

## Shutdown and interruption callbacks

Kernel users have the ability to interrupt, restart and stop the kernel.
What does it mean?

**Interruption** just interrupts current cell execution. For example,
if the user has run very long or infinite computation, they might want
to interrupt it. After interruption kernel state remains untouched.

**Stopping** kernel means just stopping the session. Kernel process is aborted,
kernel state is aborted too.

**Restarting** kernel means stopping followed by starting. Kernel process
is aborted and then started again, kernel state is also cleared.

Sometimes you may want to write a callback on kernel interruption or shutdown (that happens on stopping or restarting). And you can, using descriptor or Kotlin API.

In [25]:
// Interruption callbacks aren't supported by descriptor API
loadLibraryDescriptor("""
    {
        "shutdown": [
            "println(\"Bye!\")"
        ]
    }
""")

In [26]:
var interruptionCount = 0
USE {
    onShutdown {
        println("Bye! (2)")
    }

    onInterrupt {
        ++interruptionCount
    }
}

In [ ]:
while(true) {}

Shutdown callbacks can be managed by corresponding processor.
API for extensions management will be most likely reworked in the future
versions of kernel.

In [ ]:
interruptionCount

In [ ]:
notebook.shutdownExecutionsProcessor.unregisterAll()
notebook.interruptionCallbacksProcessor.unregisterAll()

## Color scheme change callbacks

Sometimes you may want to do something when the user changes current
color scheme in their client. For example, you might want to change color
scheme for newly created outputs or even update already created ones.
You definitely can, but only in Kotlin API, and only if the user uses Kotlin Notebook plugin

In [ ]:
var isDark = notebook.currentColorScheme == ColorScheme.DARK
USE {
    println("Dark? - $isDark")

    onColorSchemeChange { colorScheme ->
        isDark = colorScheme == ColorScheme.DARK
        println("Color scheme is changed")
    }
}

In [ ]:
isDark

In [ ]:
notebook.colorSchemeChangeCallbacksProcessor.unregisterAll()

## Rendering

Imagine you're developing a model for TicTacToe game and want to experiment with it in the notebook.
You want to have some visual representation of the model. Now all what you have is a following
string representation of the game state:

In [ ]:
val gameState = "x_o|_x_|___"
gameState

First thing you might want to do is to create a matrix containing the state in easier-to-use form

In [ ]:
enum class TTTCell {
    X, O, Empty
}

fun getGameState(state: String) = state.split('|')
    .map { line ->
        line.map {
            when (it) {
                'x' -> TTTCell.X
                'o' -> TTTCell.O
                '_' -> TTTCell.Empty
                else -> error("Not possible")
            }
        }
    }

In [ ]:
val stateMatrix = getGameState(gameState)
stateMatrix

In [ ]:
class TTTState(val stateMatrix: List<List<TTTCell>>)

In [ ]:
val state = TTTState(stateMatrix)

To render it, we may add a text renderer. Text renderers are supported only by Kotlin API and can only
render values to string. Their benefit is that they can easily stack together: string are addable.

In [ ]:
USE {
    addTextRenderer(TextRenderer { _, value ->
        (value as? TTTState)?.let { state ->
            state.stateMatrix.joinToString("\n") { line ->
                line.joinToString(" ") {
                    when (it) {
                        TTTCell.X -> "X"
                        TTTCell.O -> "O"
                        TTTCell.Empty -> "_"
                    }
                }
            }
        }
    })
}

In [ ]:
state

It's much better now, but we still might wish some graphic representation of the state.
Let's create a couple of renderers. Renderers accept some values and return some other
values. Rendering logic is described in [kernel documentation](https://github.com/Kotlin/kotlin-jupyter/blob/master/docs/README.md#rendering). Here, we register created renderers with `RenderersProcessor` and specify priority for them to ensure the order of  probation

In [ ]:
val htmlStateRenderer = createRenderer<TTTState> {
    val style = """
        border: 1px solid black;border-collapse: collapse;
    """.trimIndent()
    val html =
        it.stateMatrix.joinToString("\n", prefix = "<div><table style='$style'>", postfix = "</table></div>") { line ->
            line.joinToString("", prefix = "<tr style='$style'>", postfix = "</tr>") { cell ->
                buildString {
                    append("<td style='$style'>")
                    append(
                        when (cell) {
                            TTTCell.X -> "X"
                            TTTCell.O -> "O"
                            TTTCell.Empty -> "\u200D"
                        }
                    )
                    append("</td>")
                }
            }
        }
    mimeResult(
        MimeTypes.HTML to html,
        MimeTypes.PLAIN_TEXT to it.stateMatrix.toString()
    )
}
notebook.renderersProcessor.registerWithoutOptimizing(htmlStateRenderer, ProcessingPriority.DEFAULT + 5)

In [ ]:
state

The following renderer will render not to `DisplayResult`, but to a `BufferedImage`.
Kernel knows how to render `BufferedImage`, that's because it works fine.

In [ ]:
import java.awt.Color
import java.awt.geom.Ellipse2D
import java.awt.image.BufferedImage

@OptIn(ExperimentalStdlibApi::class)
val imageStateRenderer = createRenderer<TTTState> { state ->
    val cellSize = 100
    val padding = 5
    val mtx = state.stateMatrix
    val rowsCount = mtx.size
    val colsCount = mtx.first().size
    val width = 2 * padding + cellSize * colsCount
    val height = 2 * padding + cellSize * rowsCount

    val img = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
    val g = img.createGraphics()
    g.color = Color.BLACK

    val figPadding = cellSize / 10
    fun drawCell(xOffset: Int, yOffset: Int, cell: TTTCell) {
        when (cell) {
            TTTCell.X -> {
                g.drawLine(
                    xOffset + figPadding,
                    yOffset + figPadding,
                    xOffset + cellSize - figPadding,
                    yOffset + cellSize - figPadding
                )
                g.drawLine(
                    xOffset + cellSize - figPadding,
                    yOffset + figPadding,
                    xOffset + figPadding,
                    yOffset + cellSize - figPadding
                )
            }
            TTTCell.O -> {
                g.draw(Ellipse2D.Double(
                    (xOffset + figPadding).toDouble(),
                    (yOffset + figPadding).toDouble(),
                    (cellSize - 2 * figPadding).toDouble(),
                    (cellSize - 2 * figPadding).toDouble(),
                ))
            }

            TTTCell.Empty -> {}
        }
    }

    fun drawGrid() {
        for (i in 0..rowsCount) {
            g.drawLine(
                padding,
                padding + i * cellSize,
                padding + colsCount * cellSize,
                padding + i * cellSize,
            )
        }
        for (j in 0..colsCount) {
            g.drawLine(
                padding + j * cellSize,
                padding,
                padding + j * cellSize,
                padding + rowsCount * cellSize,
            )
        }
    }

    drawGrid()
    for (i in 0..< rowsCount) {
        for (j in 0..< colsCount) {
            drawCell(padding + j * cellSize, padding + i * cellSize, mtx[i][j])
        }
    }

    img
}

notebook.renderersProcessor.registerWithoutOptimizing(imageStateRenderer, ProcessingPriority.HIGHER)

In [ ]:
state

Renderer for `BufferedImage` is already registered!

In [ ]:
USE {
    render<TTTState> { it }
}

In [ ]:
notebook.renderersProcessor.registeredRenderers()
    .map { it.renderer }
    .filter { (it as? RendererTypeHandler)?.acceptsType(BufferedImage::class) == true }

You can add a renderer in the library descriptor as well, but functionality isn't
so rich there. In fact, you can only write code-based renderers, and renderers are
matched by **exact** FQN match. Subtypes aren't accepted.

In [ ]:
val dollar = "$"
val dollarIt = "\$it"

class Customer2(val name: String, val age: Int)

loadLibraryDescriptor(
    """
        {
            "renderers": {
                "${Customer2::class.qualifiedName}": "\"Customer $dollar{${dollarIt}.name} aged $dollar{${dollarIt}.age}\""
            }
        }
    """
)

In [ ]:
Customer2("Ada", 42)

One more thing that you can render are exceptions! Let's add a renderer for `NullPointerException`

In [ ]:
USE {
    renderThrowable<NullPointerException> {
        HTML("<b>Kotlin is a null-safe language</b>")
    }
}

In [ ]:
System.getProperty("some_unexistent_property")!!

You can also render having current execution host in the context

In [ ]:
USE {
    fun TTTCell.toHTML() = HTML("<TODO>")

    renderWithHost<TTTCell> { executionHost, value ->
        executionHost.execute {
            display(value.toHTML(), "12")
            display(textResult(value.toString()), "")
        }
    }
}

## Code preprocessing

Kotlin kernel gives you an ability to preprocess the code before execution.
One of the examples of preprocessing is magics. They are preprocessed with a built-in processor.
But sometimes you have to write your own one.

For now, you can't run suspend functions directly in the notebook cells. Let's
try to improve the situation writing a workaround for this problem. We start
with creating a function

In [ ]:
%use coroutines

In [ ]:
private class Decl(val varMod: String, val name: String)

fun generateRunBlocking(code: String): String {
    val declarationRegex = """^(va[lr]) +(\w+) """.toRegex(RegexOption.MULTILINE)
    val matches = declarationRegex.findAll(code).toList()
    val newCode = code.toCharArray()
    val declarations = matches.map { m ->
        val modGroup = m.groups[1]!!
        val modifier = modGroup.value
        for (i in modGroup.range) {
            newCode[i] = ' '
        }
        val varName = m.groups[2]!!.value
        Decl(modifier, varName)
    }
    return buildString {
        for (decl in declarations) {
            append("var ${decl.name}: Any? = null\n")
        }
        append("runBlocking {\n")
        append(String(newCode))
        append("\n}")
    }
}

Let's test it

In [ ]:
generateRunBlocking("""
    val xyz = 42
    var t = 12
""".trimIndent())

In [ ]:
USE {
    preprocessCode { code -> generateRunBlocking(code) }
}

In [ ]:
suspend fun susF(): Int = 42

val a = susF()

In [ ]:
a

In [ ]:
val processors = notebook.codePreprocessorsProcessor.registeredExtensions().toList()
println(processors)

In [ ]:
notebook.codePreprocessorsProcessor.unregister((processors!! as List<CodePreprocessor>)[0])

In [ ]:
val processors2 = notebook.codePreprocessorsProcessor.registeredExtensions().toList()
println(processors2)

## Reflection- and compile-based postprocessing

As we discussed above, there is possibility to add after-evaluation callbacks.
There is a possibility to consider script instance as well. Script instance
contains all top-level variables and functions you define as members. Let's take a look

In [ ]:
import kotlin.reflect.KProperty
import kotlin.reflect.KProperty1
import kotlin.reflect.full.memberFunctions
import kotlin.reflect.full.memberProperties

USE {
    afterCellExecution { snippetInstance, _ ->
        val kClass = snippetInstance::class
        println(
            """
Variables:
${
    kClass.memberProperties.joinToString("\n") { prop ->
        "${prop.name} = ${
            (prop as KProperty1<Any, Any?>).get(
                snippetInstance
            )
        }"
    }
}

Functions:
${kClass.memberFunctions.joinToString("\n") { f -> f.name }}
        """
        )
    }
}

In [ ]:
val myVar1 = 42
val myVar2 = "str"

fun myFun1() { println("fun called") }

In [ ]:
notebook.afterCellExecutionsProcessor.unregisterAll()

Let's now try to do something compile-time based. We start with file annotations processing

In [ ]:
USE {
    onFileAnnotation<Suppress> {
        val myAnno = it.first() as Suppress
        println("Warnings suppressed: " + myAnno.names.toList())
    }
}

In [ ]:
@file:Suppress("UNUSED_PARAMETER")

fun u(unusedArg: Int) = 44

We can also handle annotations for classes. They are runtime-based.

In [ ]:
annotation class MyAnnotation

In [ ]:
USE {
    onClassAnnotation<MyAnnotation> { println("Annotated classes: $it") }
}

In [ ]:
@MyAnnotation
class MyClass

class MyClass2

@MyAnnotation
interface MyInterface3

3 + 3

In [ ]:
class MyData

USE {
    onVariable<MyData> { t, kProperty ->
        execute("val gen_${kProperty.name} = 1")
    }
//    updateVariable<MyData> { t, kProperty ->
//        execute("")
//
//    }
}

In [ ]:
val d = MyData()

In [ ]:
gen_d

## Variables reports

In [ ]:
notebook.variablesReport

In [ ]:
USE {
    markVariableInternal { prop ->
        prop.name.startsWith("___")
    }
}

In [ ]:
val ___b = 8
val b = 9

In [ ]:
notebook.variablesReport

## Auxiliary

In [ ]:
// You can set minimal supported version of kernel
USE {
    setMinimalKernelVersion("0.11.0.365")
}

In [ ]:
USE {
    setDescription("Very good library")
    setWebsite("https://cool.site")
}